System/RTETools.h

```diff
@@ -38,37 +40,74 @@ namespace RTE {
 
 #pragma region Random Numbers
 	/// <summary>
-	/// Seeds the rand with the current runtime time.
+	/// Seed the mt19937 random number generator.
+	/// mt19937 is the standard mersenne_twister_engine.
 	/// </summary>
-	void SeedRand();
+	void SeedRNG();
 
 	/// <summary>
-	/// A good rand function that return a float between 0.0 and 0.999.
+	/// Seed the mt19937 random number generator.
+	/// mt19937 is the standard mersenne_twister_engine.
 	/// </summary>
-	/// <returns>Random number.</returns>
+	/// <param name="seed">Seed for the random number generator.</param>
+	void SeedRNG(unsigned int seed);
+
+	/// <summary>
+	/// Generate a pre-seed for the mt19937 random number generator.
+	/// The pre-seed is used in SeedRNG with seed = 0 to create a good seed sequence for the RNG.
+	/// </summary>
+	/// <returns>A pre-seed for the mt19937 random number generator.</returns>
+	std::array<int, 624> GeneratePreSeed();
+
+	/// <summary>
+	/// Uniformly distributed random double in the range [0, 1].
+	/// </summary>
+	/// <returns>Uniformly distributed random double in the range [0, 1].</returns>
 	double PosRand();
 
 	/// <summary>
-	/// A good rand function that returns a floating point value between -1.0 and 1.0, both inclusive.
+	/// Uniformly distributed random float in the range [-1, 1].
 	/// </summary>
-	/// <returns>Random number.</returns>
-	double NormalRand();
+	/// <returns>Uniformly distributed random float in the range [-1, 1].</returns>
+	float NormalRand();
 
 	/// <summary>
-	/// A good rand function that returns a floating point value between two given thresholds, the min being inclusive, but the max not.
+	/// Template function which returns a Uniformly distributed random number in the range [0, 1].
 	/// </summary>
-	/// <param name="min">Minimum value this can return.</param>
-	/// <param name="max">Maximum value this can return.</param>
-	/// <returns>Random number between limits.</returns>
-	double RangeRand(float min, float max);
+	/// <param name="min">Lower boundary of the range to pick a number from.</param>
+	/// <param name="max">Upper boundary of the range to pick a number from.</param>
+	/// <returns>Uniformly distributed random number in the range [0, 1].</returns>
+	template <typename floatType = float>
+	typename std::enable_if<std::is_floating_point<floatType>::value, floatType>::type RandomNum() {
+		return std::uniform_real_distribution<floatType>(floatType(0.0), std::nextafter(floatType(1.0), std::numeric_limits<floatType>::max()))(RTETools_RNG);
+	}
+
+	template <typename intType>
```

Comment:
This should default to `int` otherwise calling it without specifying `<int>` appears to return an `unsigned int` result.
Not sure if it actually does but I get a lint warning if there's no default.
See comment about not specifying `<int>` in `Deployment.h`.

Reply:
Here it gets a little complicated.
The default type for the template is already float. It may look like a separate template but it's the same one.
I set the default to float because it's the most often-used type.

Reply:
Yeah ok, looks like I dumd here a lil bit.
I was getting a lint warning because the variable was unsigned.
So never mind this.